File: src/algorithms/is-monotonic/is-monotonic.ts
/**
 * An array is monotonic if it is either monotone increasing or monotone decreasing.
 *
 * An array nums is monotone increasing if for all i <= j, nums[i] <= nums[j].
 *
 * An array nums is monotone decreasing if for all i <= j, nums[i] >= nums[j].
 *
 * Given an integer array nums, return true if the given array is monotonic, or false otherwise.
 */
export function isMonotonic(nums: number[]): boolean {
  return (
    nums.every((element, index) => index === 0 || element <= nums[index - 1]) ||
    nums.every((element, index) => index === 0 || element >= nums[index - 1])
  )
}

export function isMonotonic1(nums: number[]): boolean {
  let inc = true
  let dec = true

  for (let i = 0; i < nums.length - 1; i++) {
    // check if the number in the arr is not increasing
    if (nums[i] > nums[i + 1]) {
      inc = false
    }
  }

  for (let i = 0; i < nums.length - 1; i++) {
    // check if the number in the arr in not decreasing
    if (nums[i] < nums[i + 1]) {
      dec = false
    }
  }

  return inc || dec
}

File: src/algorithms/arrays/is-monotonic/is_monotonic.py
"""
Is Monotonic

An array is monotonic if it is either monotone increasing or monotone decreasing.

An array nums is monotone increasing if for all i <= j, nums[i] <= nums[j].

An array nums is monotone decreasing if for all i <= j, nums[i] >= nums[j].

Given an integer array nums, return true if the given array is monotonic, or false otherwise.

"""

from typing import List


class Solution:
    def isMonotonic(self, nums: List[int]) -> bool:
        return True if (nums == sorted(nums) or nums == sorted(nums, reverse=True)) else False


class AnotherSolution:
    def isMonotonic(self, nums: List[int]) -> bool:
        return all(nums[i] <= nums[i+1] for i in range(len(nums)-1)) or all(nums[i] >= nums[i+1] for i in range(len(nums)-1))
